### DISCUSSION TITLE
Add capability for using stale cache while JWKS endpoint is down

### DISCUSSION THREAD

**Discussion by flaviostutz** [panva/jose#415] [Ideas]
Sometimes the JWKS endpoint may be down. In these cases, it would be useful if jose could still use a stale version of the cached keys even after cacheMaxAge has passed to avoid outages in a service.

**Comment by KeithRichardsonKR** [comment]
I think this is a very important feature.  It's not good for a network issue or outage on the JWKS endpoint to take down the JWT-secured API, so it leaves us with two choices:

1. Set cacheMaxAge to a very log time/infinite so that the cache never expires (a new kid will refresh the cache)
2. Keep a cacheMaxAge, but potentially have an outage if JWKS is ever down

Would it be reasonable to update getKey to catch errors on this.Reload() and continue if there was an error in reloading the cache?
Code link as of time of post: https://github.com/panva/jose/blob/f009cc16d966774a64a6cfff0facf5515bf64411/src/jwks/remote.ts#L117

Maybe there can be config of whether to allow using a stale cache?  Alternatively, there could be a staleMaxAge as well, so you could have your config set like this:

cacheMaxAge: 1 day
staleMaxAge: 2 days

So it will try to refresh after 1 day, but if there's errors on reload() it will use the stale cache up to another day, after which it will start returning errors like it currently does.  The staleMaxAge could default to be the same as cacheMaxAge if not set.

Would you be open to a pull request for this feature?

**Comment by KeithRichardsonKR** [comment]
@panva, my proposal would be to create a `cacheStaleMaxAge` config setting.  If an error occurs during `this.reload()` and `cacheStaleMaxAge` milliseconds have not elapsed since the cache was last loaded, it would use the existing cached keys.  If there was an error and `cacheStaleMaxAge` milliseconds have elapsed, it would return the error as normal.

Would you be willing to consider a pull request with this feature if I submitted it?  Thanks!

**Reply by panva** [reply]
I don't quite get the point. You already control accepting "stale" with the existing cache option. 

If a rotation is detected and the jwks endpoint is down the token which needs keys fetched still wont validate while others will be fine.
